File: 20-valid-parentheses/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    var stack =[];
    if (s.length==0) return false;
    if ((s.length ==0)||s.length==1) return false;
    for(var i=0; i<s.length; i++){
        if(s[i]=="}"||s[i]==")"||s[i]=="]"){
            if(stack.length==0) return false;
            if(s[i]=="}" && stack[stack.length-1] != "{") return false;
            if(s[i]==")" && stack[stack.length-1] != "(") return false;
            if(s[i]=="]" && stack[stack.length-1] != "[") return false;
            stack.pop();
        }else{
            stack.push(s[i]);
        }
    }
    if(stack.length==0) return true;
    return false;
};

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack <Character> stack = new Stack<Character>();
        
        Map <Character, Character> hm = new HashMap<Character, Character>();
        hm.put(')','(');
        hm.put('}','{');
        hm.put(']','[');
        
        for(int i=0; i<s.length(); i++){
            char tmp = s.charAt(i);
            if(!hm.containsKey(tmp)){
                stack.push(tmp);
            }else{
                char ob = hm.get(tmp);
                if(stack.size()==0 || stack.pop()!=ob) return false;
            }
        }
        if(stack.size()>0) return false;
        return true;
    }
}
